File: src/com/kedzie/vbox/metrics/MetricView.java
package com.kedzie.vbox.metrics;

import java.util.Map;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.res.TypedArray;
import android.support.v4.content.LocalBroadcastManager;
import android.util.AttributeSet;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.kedzie.vbox.R;
import com.kedzie.vbox.VBoxApplication;

public class MetricView extends LinearLayout {

	private TextView _titleTextView;
	private MetricRenderer _renderer;
	private LinearLayout _metricNames;
	private String _header;
	private BroadcastReceiver _receiver = new BroadcastReceiver() {
        @Override
        public void onReceive(Context context, Intent intent) {
            if(intent.getAction().equals(MetricActivity.ACTION_METRIC_QUERY)) {
//                _renderer.setQuery(q);
            }
        }
    };
	
    public MetricView(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }
    
    public MetricView(Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
        TypedArray a = context.getTheme().obtainStyledAttributes(attrs, R.styleable.MetricView, 0, 0);
        try {
            _header = a.getString(R.styleable.MetricView_header);
            int bgColor = a.getColor(R.styleable.MetricView_backgroundColor, android.R.color.white);
            int gridColor = a.getColor(R.styleable.MetricView_gridColor, android.R.color.black);
            int textColor = a.getColor(R.styleable.MetricView_textColor, android.R.color.black);
            int borderColor = a.getColor(R.styleable.MetricView_borderColor, android.R.color.holo_blue_dark);
            createView(bgColor, gridColor, textColor, borderColor);
        } finally {
            a.recycle();
        }
    }
	
	public MetricView(Context context, String header) {
	    super(context);
	    _header=header;
	    createView(0xFFFFFFFF, android.R.color.black, android.R.color.black, android.R.color.holo_blue_dark);
	}
	
	public void createView(int bgColor, int gridColor, int textColor, int borderColor) {
	    setOrientation(LinearLayout.VERTICAL);
	    _titleTextView = new TextView(getContext());
        _titleTextView.setText(_header);
//        _titleTextView.setTextAppearance(getContext(), android.R.style.TextAppearance_Large);
        addView(_titleTextView, new LinearLayout.LayoutParams(LinearLayout.LayoutParams.MATCH_PARENT, LinearLayout.LayoutParams.WRAP_CONTENT));
        _renderer = new MetricRenderer(getContext(), bgColor, gridColor, textColor, borderColor);
        LinearLayout.LayoutParams p = new LinearLayout.LayoutParams(LinearLayout.LayoutParams.MATCH_PARENT, LinearLayout.LayoutParams.WRAP_CONTENT);
        p.weight=1;
        addView(_renderer, p);
        _metricNames = new LinearLayout(getContext());
        _metricNames.setOrientation(LinearLayout.HORIZONTAL);
        addView(_metricNames, new LinearLayout.LayoutParams(LinearLayout.LayoutParams.MATCH_PARENT, LinearLayout.LayoutParams.WRAP_CONTENT));
        LocalBroadcastManager.getInstance(getContext()).registerReceiver(_receiver, new IntentFilter(MetricActivity.ACTION_METRIC_QUERY));
	}

	public void init(int max, String []metrics) {
		_renderer.init(max, metrics);
		LinearLayout.LayoutParams p = new LinearLayout.LayoutParams(LinearLayout.LayoutParams.WRAP_CONTENT, LinearLayout.LayoutParams.WRAP_CONTENT);
		for(String m : metrics) {
			TextView textView = new TextView(getContext());
			textView.setText(m);
			textView.setTextColor(VBoxApplication.getInstance().getColor(getContext(), m.replace('/', '_')));
			textView.setPadding(0,2,8,0);
			_metricNames.addView(textView, p);
		}
	}
	
	public void close() {
	    LocalBroadcastManager.getInstance(getContext()).unregisterReceiver(_receiver);
	}
	
	public void setMetricPrefs(int count, int period) {
		_renderer.setMetricPrefs(count, period);
	}

	public void setQueries(Map<String,MetricQuery> q) {
		_renderer.setQuery(q);
	}
	
	public String getHeader() {
	    return _header;
	}
}

File: vbox-app/src/main/java/com/kedzie/vbox/metrics/MetricView.kt
package com.kedzie.vbox.metrics

import android.content.Context
import android.content.res.TypedArray
import android.util.AttributeSet
import android.widget.LinearLayout
import android.widget.TextView

import androidx.appcompat.widget.AppCompatTextView

import com.kedzie.vbox.R
import com.kedzie.vbox.VBoxApplication

class MetricView @JvmOverloads constructor(context: Context, attrs: AttributeSet, defStyle: Int = 0) : LinearLayout(context, attrs) {

    var header: String

    private lateinit var renderer: MetricRenderer
    private lateinit var metricNames: LinearLayout

    init {
        val a = context.theme.obtainStyledAttributes(attrs, R.styleable.MetricView, 0, R.style.MetricView)
        try {
            header = a.getString(R.styleable.MetricView_header)
            val bgColor = a.getColor(R.styleable.MetricView_backgroundColor, 0)
            val gridColor = a.getColor(R.styleable.MetricView_gridColor, 0)
            val textColor = a.getColor(R.styleable.MetricView_textColor, 0)
            val borderColor = a.getColor(R.styleable.MetricView_borderColor, 0)
            createView(bgColor, gridColor, textColor, borderColor)
        } finally {
            a.recycle()
        }
    }

    fun createView(bgColor: Int, gridColor: Int, textColor: Int, borderColor: Int) {
        orientation = VERTICAL
        val titleTextView = AppCompatTextView(context)
        titleTextView.text = header
        addView(titleTextView, LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.WRAP_CONTENT))
        renderer = MetricRenderer(context, bgColor, gridColor, textColor, borderColor)
        val p = LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.WRAP_CONTENT)
        p.weight = 1f
        addView(renderer, p)
        metricNames = LinearLayout(context)
        metricNames.orientation = HORIZONTAL
        addView(metricNames,LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.WRAP_CONTENT))
    }

    fun init(max: Int, metrics: Array<String>) {
        renderer.init(max, metrics)
        val p = LayoutParams(LayoutParams.WRAP_CONTENT, LayoutParams.WRAP_CONTENT)
        for (m in metrics) {
            val textView = AppCompatTextView(context)
            textView.text = m
            textView.setTextColor(VBoxApplication.getInstance().getColor(context, m.replace('/', '_')))
            textView.setPadding(0, 2, 8, 0)
            metricNames.addView(textView, p)
        }
    }

    fun setMetricPrefs(count: Int, period: Int) {
        renderer.setMetricPrefs(count, period)
    }

    fun setQueries(q: Map<String, MetricQuery>) {
        renderer.setQuery(q)
    }
}
